source/engine/graphics/RenderGraph.cpp

```diff
@@ -0,0 +1,211 @@
+      m_gpuAllocator{gpuAllocator},
+      m_descriptorSets{descriptorSets},
+      m_dimensions{dimensions}
+{
+    Add(CreateLitPass(m_device, m_gpuAllocator, m_gpuOptions, m_swapchain, m_dimensions));
+
+    #ifdef NC_EDITOR_ENABLED
+    RegisterTechnique<WireframeTechnique>(LitPassId);
+    #endif
+
+    RegisterTechnique<EnvironmentTechnique>(LitPassId);
+    RegisterTechnique<PbrTechnique>(LitPassId);
+    RegisterTechnique<ParticleTechnique>(LitPassId);
+    RegisterTechnique<UiTechnique>(LitPassId);
+}
+
+RenderGraph::~RenderGraph() noexcept
+{
+    m_renderPasses.clear();
+}
+
+void RenderGraph::Execute(PerFrameGpuContext *currentFrame, const PerFrameRenderState &frameData, const MeshStorage &meshStorage, uint32_t frameBufferIndex, Vector2 dimensions) const
+{
+    OPTICK_CATEGORY("RenderGraph::Execute", Optick::Category::Rendering);
+
+    const auto cmd = currentFrame->CommandBuffer();
+    SetViewportAndScissor(cmd, dimensions);
+    BindMeshBuffers(cmd, meshStorage.GetVertexData(), meshStorage.GetIndexData());
+
+    auto bufferIndex = 0u;
+    for (auto &renderPass : m_renderPasses)
+    {
+        BeginRenderPass(renderPass.get(), cmd, frameBufferIndex);
+        for (const auto &technique : renderPass->techniques)
+        {
+            if (!technique->CanBind(frameData)) continue;
+
+            technique->Bind(cmd);
+
+            if (!technique->CanRecord(frameData)) continue;
+
+            technique->Record(cmd, frameData);
+        }
+        EndRenderPass(cmd);
+    }
+    cmd->end();
+}
+
+RenderPass* RenderGraph::Acquire(const std::string& uid)
+{
+    const auto renderPassPos = std::ranges::find_if(m_renderPasses, [uid](auto& renderPass) { return (renderPass->uid == uid); });
+    if (renderPassPos == m_renderPasses.end()) throw NcError("RenderGraph::Acquire - Render pass does not exist.");
+    return (*renderPassPos).get();
+}
+
+void RenderGraph::Add(std::unique_ptr<RenderPass> renderPass)
+{
+    m_renderPasses.push_back(std::move(renderPass));
+    std::ranges::sort(m_renderPasses, [](const auto &renderPassA, const auto &renderPassB)
+    {
+        return renderPassA->priority < renderPassB->priority;
+    });
+}
+
+void RenderGraph::Remove(const std::string& uid)
+{
+    std::erase_if(m_renderPasses, [uid](const auto &pass)
+    {
+        return pass->uid == uid;
+    });
+}
+
+void RenderGraph::RegisterShadowMappingTechnique(const std::string& uid, uint32_t shadowCasterIndex)
+{
+    UnregisterTechnique<ShadowMappingTechnique>(uid);
+    auto* renderPass = Acquire(uid);
+    renderPass->techniques.push_back(std::make_unique<ShadowMappingTechnique>(m_device, m_gpuOptions, m_descriptorSets, renderPass->renderPass.get(), shadowCasterIndex));
+}
+
+void RenderGraph::Resize(Swapchain* swapchain, const Vector2& dimensions)
+{
+    m_renderPasses.clear();
+    m_dimensions = dimensions;
+    m_swapchain = swapchain;
+    Add(CreateLitPass(m_device, m_gpuAllocator, m_gpuOptions, m_swapchain, m_dimensions));
```

Comment:
Since the stuff here and below is also done in the c'tor, I wonder if it should be factored out.